### PR TITLE
[ot-math-table] fix a bug in MathKern::get_value()

### PR DESCRIPTION
**Pull request by lie-yan**
Fixes harfbuzz/harfbuzz#4542

### CODE REVIEW

**Comment by lie-yan** [comment]
Fixes harfbuzz/harfbuzz#4542

**Comment by lie-yan** [comment]
Fixes harfbuzz/harfbuzz#4542

File: src/hb-ot-math-table.hh [hunk 1]
```diff
@@ -344,26 +344,32 @@ struct MathKern
     const MathValueRecord* kernValue = mathValueRecordsZ.arrayZ + heightCount;
     int sign = font->y_scale < 0 ? -1 : +1;
 
-    /* The description of the MathKern table is a ambiguous, but interpreting
-     * "between the two heights found at those indexes" for 0 < i < len as
-     *
-     *   correctionHeight[i-1] < correction_height <= correctionHeight[i]
-     *
-     * makes the result consistent with the limit cases and we can just use the
-     * binary search algorithm of std::upper_bound:
```
Comment (fred-wang) [inline src/hb-ot-math-table.hh:353]: With your changes, isn't the loop below actually exactly doing the same as `std::upper_bound` (see https://en.cppreference.com/w/cpp/algorithm/upper_bound#Possible_implementation)?

If so, I think it's not necessary to add a comment with the explanation of the algorithm below. Instead, I'd keep the reference to `std::upper_bound` and explain that this matches the spec, including for the boundary cases.

Comment (lie-yan) [reply src/hb-ot-math-table.hh:353]: You are right. That's just what `std::upper_bound()` does.
I have changed the comment as you suggested.

**Review by fred-wang — COMMENTED** [review]
Thanks. This LGTM, but I'm not a maintainer harfbuzz so I'll defer official review to someone else.

**Comment by behdad** [comment]
I wonder if we could use our internal bsearch?

**Comment by lie-yan** [comment]
> I wonder if we could use our internal bsearch?

I believe we can. We can use the following code snippet to compute index `i`. 
```cpp
unsigned pos;
unsigned i = hb_bsearch_impl(&pos, key, base, nmemb, stride, compar) ? pos + 1 : pos;
```

For `key` in the array, `pos` points to the element that is equal to `key`.
For `key` not in the array, we can verify that the algorithm is equivalent to the following code snippet, which is exactly [`std::upper_bound()`](https://en.cppreference.com/w/cpp/algorithm/upper_bound#Version_1) except for the computation of `step`. We can prove that they are equivalent. 

```cpp
int min = 0, count = nmemb;
// Define:
//   max := min + count - 1
// Assume:
//   array[-1] = -infty
//   array[nmemb]= +infty
// Loop invariant:
//   0 <= min + count <= nmemb
//   array[min-1] <= key < array[min + count]
// Termination:
//   count is decreased in each iteration and reaches 0 on loop end
while (count > 0) 
{
  int step = (count - 1) / 2;
  int mid = min + step;
  
  if (!(key < array[mid])) {
    min = mid + 1;
    count -= step + 1;
  }
  else {
    count = step;
  }
}
pos = min;
```

**Comment by behdad** [comment]
Thanks. Would be nice to do it.

**Comment by lie-yan** [comment]
I will create a pull request.

**Pull request merged**